### DISCUSSION TITLE
Why is the return type from `upsert` Partial?

### DISCUSSION THREAD

**Discussion by eXamadeus** [tywalch/electrodb#349] [Q&A]
First, I love this library! It's amazing. I tried others, and there is no comparison.

So, to my question... I've been trying to use the `upsert` mechanism for a few things, and I'm confused about the `upsert` return type. Whenever I make the `upsert` call, the data type is `Partial`, which confuses me. Shouldn't this ALWAYS return a result like `create`?

Here is a [playground link](https://electrodb.fun/?#code/JYWwDg9gTgLgBAbzgUQHY2DAnnAvnAMyghDgCIBTAGwoGMZiATAIzIG4AoD2iVAZ3gwAhnwDWfOAF44qCgHcU6TFgAUHOInUa4ICI2oAuTdu0Ul2I2WFi+ZADRaTANwpQ+wXpYCM9x9r6uTsC0FJbWokJgYGR+uA4mQjAMwMwArjAUfEYIfhoZQiDZudrYYKHkAlDAqADmviYmUBQAjqnATYxGDKkUxXHFqQFQRQ0lWGWWldV18aNwTa3tFJ1w3b2j-aMYMDQjc6XlZFO19XMLbR1dUD2zJpva9xrV+gAemXsmYMQAVnQwWcY5mBRB85gRgNQVmRgac5nAeOAIO4MkYANpkQauMgAXWKGkeDTEoNG4Mhk1EsLmCMgyPK6O2NBxeLwfViWnuSGEzF25CwEFSUAA+lyaILUAUKGQWQBKTgccJ8AB0qTAQxgKgQ+UKcAA5AAJABKOrsq0wPJ1AGEABZ0URwCggITAKjGuCY4a6gDKGRcOtw0sVNQgKgDMBtqBUGrgjESQjw0qkAD5ATx+BAaIqqBAaioY8JFVqEwB6IurcYUOAQAjR2MFigFODACQ6441OAAHzdqH04NkjB1HH9QA) to show what I'm talking about.

This may be the expected behavior since I'm new to DynamoDB and expect it to work like the `create` method. Is it partial because it doesn't perform a query (just the update), and thus, it can't know what the entire object looks like? If so, that makes sense. It might help to outline that in the documentation.

**Comment by tywalch** [comment]
Let me look into this, I agree at first glance 👍

**Reply by tywalch** [reply]
@eXamadeus 👋 

I looked into this and found why the return typing is not the same as `create` or `put`. The `create` and `put` have the response type they do because the values you pass will be the values you write; both methods use DynamoDB `put` under the hood so they will both overwrite any existing value. The `upsert` method, while having a similar interface to `create` and `put`, does not necessarily write the same values that it receives. Two notable examples:
1. The `upsert` operation can also perform an `update` which means even if the method returned the values provided, it would not be reflective of the item in the database. 
2. The values passed to the `add`, `subtract`, `append` and `ifNotExists` methods on `upsert` can result in a different value in the database than what was provided. Readonly values get automatically added as `ifNotExists`, so these values could also be different.

This is why `upsert` does not have the same response by default as `create` and `put`. That said, you can easily get the same response by using the execution option `.go({ response: 'all_new' })` to return the stored item as it exists after write 👍
